docs/nb/bulk_validation_referential_integrity_check.ipynb

```diff
@@ -190,78 +305,56 @@
-    "    assert len(collection_name_to_class_names[name]) == 1"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "id": "5ed95ee0-03b7-4dff-80e7-92a2b24bccf4",
-   "metadata": {},
-   "source": [
-    "Define a helper function that takes a class instance and returns a list of the names of its own class and its ancestor classes."
+    "# Any ancestor of a document class is a document-referenceable range, i.e., a valid range of a document-reference-ranged slot.\n",
+    "document_referenceable_ranges = set(chain.from_iterable(schema_view.class_ancestors(cls_name) for cls_name in document_class_names))\n",
+    "\n",
+    "document_reference_ranged_slots = defaultdict(list)\n",
+    "for cls_name, slot_map in cls_slot_map.items():\n",
+    "    for slot_name, slot in slot_map.items():\n",
+    "        if str(slot.range) in document_referenceable_ranges:\n",
+    "            document_reference_ranged_slots[cls_name].append(slot_name)"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 7,
-   "id": "4470c52a-81e4-4511-b549-768c04c3b45d",
+   "execution_count": 12,
+   "id": "d253c567-533f-440f-8376-03a6e1e905cf",
    "metadata": {},
    "outputs": [],
    "source": [
-    "def class_hierarchy_as_list(obj) -> list[str]:\n",
-    "    r\"\"\"\n",
-    "    Returns a list consisting of the name of the class of the instance pass in,\n",
-    "    and the names of all of its ancestor classes.\n",
-    "\n",
-    "    TODO: Consider renaming function to be a verb; e.g. `get_class_hierarchy_as_list`.\n",
+    "def doc_cls(doc, coll_name=None):\n",
+    "    \"\"\"Return unprefixed name of document class.\n",
     "\n",
-    "    TODO: Document the purpose of the `rv` list (does not seem to be used anywhere).\n",
+    "    Try to get from doc['type'] (lopping off 'nmdc:' prefix).\n",
+    "    Else, if can unambiguously infer type given coll_name, use that.\n",
+    "    Else, return None.\n",
     "    \"\"\"\n",
-    "\n",
-    "    rv = []\n",
-    "    current_class = obj.__class__\n",
-    "    \n",
-    "    def recurse_through_bases(cls):\n",
-    "        name = cls.__name__\n",
-    "        if name == \"YAMLRoot\":  # base case\n",
-    "            return rv\n",
-    "        rv.append(name)\n",
-    "        for base in cls.__bases__:\n",
-    "            recurse_through_bases(base)  # recursive invocation\n",
-    "        return rv\n",
-    "    \n",
-    "    return recurse_through_bases(current_class)  # initial invocation"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "id": "b962e3c8-a346-49c5-8470-915f3cf9eb07",
-   "metadata": {},
-   "source": [
-    "Materialize `alldocs` collection, associating all inherited classes with document via `type` field.\n",
-    "\n",
-    "> **TODO:** Clarify the above sentence."
+    "    if 'type' in doc:\n",
+    "        return doc['type'][5:] # lop off \"nmdc:\" prefix\n",
+    "    elif coll_name and len(collection_name_to_class_names[coll_name]) == 1:\n",
```

Comment:
This doesn't need to be addressed in this PR but this else logic can be removed for berkeley, type is required and collection names now in some cases refer to abstract classes. @PeopleMakeCulture

Reply:
I created a standalone ticket about this topic. Ticket: https://github.com/microbiomedata/nmdc-runtime/issues/632